### PR TITLE
[feature/storage] Add storage for uploaded files

### PR DESCRIPTION
**Pull request by rubencm**
Checklist:

- [ ] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [ ] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-14169

### CODE REVIEW

File: phpBB/phpbb/storage/helper.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\storage;
+
+class storage
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/helper.php]: should be helper

File: phpBB/phpbb/storage/helper.php [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+{
+	/**
+	 * Eliminates useless . and .. components from specified path.
+	 *
+	 * @param string $path Path to clean
+	 *
+	 * @return string Cleaned path
+	 */
+	public function clean_path($path)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/helper.php]: should be static

File: phpBB/phpbb/storage/helper.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	}
+
+	/**
+	 * Checks if a path is absolute or not
+	 *
+	 * @param string	$path	Path to check
+	 *
+	 * @return	bool	true if the path is absolute, false otherwise
+	 */
+	public function is_absolute_path($path)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/helper.php]: should be static

File: phpBB/phpbb/storage/helper.php [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		return (isset($path[0]) && $path[0] === '/' || preg_match('#^[a-z]:[/\\\]#i', $path)) ? true : false;
+	}
+
+	/**
+	 * Try to resolve real path when PHP's realpath failes to do so
+	 *
+	 * @param string	$path
+	 * @return bool|string
+	 */
+	protected function phpbb_own_realpath($path)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/helper.php]: should be static

File: phpBB/phpbb/storage/helper.php [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	 * A wrapper for PHP's realpath
+	 *
+	 * Try to resolve realpath when PHP's realpath is not available, or
+	 * known to be buggy.
+	 *
+	 * @param string	$path	Path to resolve
+	 *
+	 * @return string	Resolved path
+	 */
+	public function realpath($path)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/helper.php]: should be static

File: phpBB/phpbb/storage/helper.php [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+	/**
+	 * Given an existing path, convert it to a path relative to a given starting path
+	 *
+	 * @param string $end_path		Absolute path of target
+	 * @param string $start_path	Absolute path where traversal begins
+	 *
+	 * @return string Path of target relative to starting path
+	 */
+	public function make_path_relative($end_path, $start_path)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/helper.php]: should be static

File: phpBB/config/default/container/services_storage.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+services:
+    storage.factory:
+        class: phpbb\storage\factory
+    storage.avatar:
+        class: phpbb\storage\storage
+    storage.attachment:
+        class: phpbb\storage\storage
+    storage.backup:
+        class: phpbb\storage\storage
```
Comment (Nicofuma) [inline phpBB/config/default/container/services_storage.yml]: removes this services definition. This part will be done later

File: phpBB/phpbb/storage/adapter/adapter_interface.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace phpbb\storage\adapter;
+
+interface adapter_interface
+{
+	public function put_contents($path, $content);
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/adapter/adapter_interface.php]: comments :)

File: phpBB/phpbb/storage/adapter/adapter_interface.php [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<?php
+
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/adapter/adapter_interface.php]: do not forgot the header file comment (will be shown by the cs check)

File: phpBB/phpbb/storage/factory.php [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\storage;
+
+class factory
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/factory.php]: removes this factory please

File: phpBB/phpbb/storage/storage.php [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+use phpbb\storage\exception\exception;
+
+class storage
+{
+	protected $adapter;
+
+	public function __construct()
+	{
+		$this->adapter = new \phpbb\storage\adapter\local();
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: `$this->adapter = new adapter\local();`

File: phpBB/phpbb/storage/storage.php [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	public function get_adapter()
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: removes this method please

File: phpBB/phpbb/storage/storage.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	{
+		return $this->adapter;
+	}
+
+	public function set_adapter($adapter, $params)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: removes this method please

File: phpBB/phpbb/storage/storage.php [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	{
+		$this->adapter = new $adapter($params);
+	}
+
+	public function put_contents($path, $content)
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/storage.php]: and comments again :)

**Comment by Nicofuma** [comment]
https://travis-ci.org/phpbb/phpbb/jobs/244502496#L545 for the coding style

**Comment by rubencm** [comment]
!set WIP

**Pull request closed without merging**